File: cxx-qt-lib/src/types/update_requester.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx::{type_id, ExternType};
use std::os::raw::{c_char, c_void};

#[cxx::bridge]
mod ffi {
    #[namespace = "rust::cxxqtlib1"]
    unsafe extern "C++" {
        include!("cxx-qt-lib/include/qt_types.h");

        type UpdateRequester = super::UpdateRequester;

        #[rust_name = "request_update"]
        fn requestUpdate(self: &UpdateRequester) -> bool;
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct UpdateRequester {
    method: *mut c_char,
    obj: *mut c_void,
}

unsafe impl ExternType for ffi::UpdateRequester {
    type Id = type_id!("rust::cxxqtlib1::UpdateRequester");
    type Kind = cxx::kind::Trivial;
}

// # Safety
//
// The underlying C++ class has been designed to be thread safe and we only
// store a pointer to it which is valid from any thread.
unsafe impl Send for UpdateRequester {}
unsafe impl Sync for UpdateRequester {}

File: cxx-qt-lib/src/update_requester.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

#include "cxx-qt-lib/include/update_requester.h"

#include <QtCore/QMetaObject>

namespace rust {
namespace cxxqtlib1 {

UpdateRequester::UpdateRequester(QPointer<QObject> obj, const char* method)
  : m_method(method)
  , m_obj(obj)
{
}

bool
UpdateRequester::requestUpdate() const
{
  if (m_obj == nullptr) {
    return false;
  }

  return QMetaObject::invokeMethod(m_obj, m_method, Qt::QueuedConnection);
}

} // namespace cxxqtlib1
} // namespace rust
